src/MassTransit/Audit/Observers/AuditPublishObserver.cs

```diff
@@ -32,7 +32,11 @@ public AuditPublishObserver(IMessageAuditStore store, ISendMetadataFactory metad
             _filter = filter;
         }
 
-        Task IPublishObserver.PrePublish<T>(PublishContext<T> context) => TaskUtil.Completed;
+        Task IPublishObserver.PrePublish<T>(PublishContext<T> context)
+        {
+            context.SentTime = DateTime.UtcNow;
```

Comment:
Why? This should be set by the transport once it's actually sent.

Reply:
I agree and it was my first implementation. But it felt weird because depending on where I put the "context.SentTime = DateTime.UtcNow;" line it could execute after the consumer had triggered and the consume audit records had been written to the database. In which case, ordering the audit records by SentTime would yield an invalid order of  events.

Therefore, I tried to find an alternative that would be less brittle to future change (this is the first time I work on this codebase).

I could change it back replacing the changes to the audit observers and add that line of code right after instantiating the BasicPublishRabbitMqSendContext (as below):

```
public async Task Send(ModelContext modelContext)
            {
                LogContext.SetCurrentIfNull(_context.LogContext);

                await _context.ConfigureTopologyPipe.Send(modelContext).ConfigureAwait(false);

                var properties = modelContext.Model.CreateBasicProperties();

                var context = new BasicPublishRabbitMqSendContext<T>(properties, _context.Exchange, _message, _cancellationToken);
                context.SentTime = DateTime.UtcNow;

                var activity = LogContext.IfEnabled(OperationName.Transport.Send)?.StartActivity(new {_context.Exchange});
                try
                {
                     ...
```

And write an unit test to check the order of events. Would you think this is a better option?